java/java.lsp.server/src/org/netbeans/modules/java/lsp/server/db/DBAddConnection.java

```diff
@@ -66,12 +70,66 @@
 @ServiceProvider(service = CodeActionsProvider.class)
 public class DBAddConnection extends CodeActionsProvider {
     public static final String DB_ADD_CONNECTION =  "db.add.connection"; // NOI18N
+    public static final String USER_ID =  "userId"; // NOI18N
+    public static final String PASSWORD =  "password"; // NOI18N
+    public static final String DRIVER =  "driver"; // NOI18N
+    public static final String DB_URL =  "url"; // NOI18N
+    public static final String SCHEMA =  "schema"; // NOI18N
+    public static final String DISPLAY_NAME =  "displayName"; // NOI18N
+    
+    private static final Set<String> COMMANDS = new HashSet<>(Arrays.asList(DB_ADD_CONNECTION));
+    
+    private final Gson gson = new Gson();
 
     @Override
     public CompletableFuture<Object> processCommand(NbCodeLanguageClient client, String command, List<Object> arguments) {
         if (!DB_ADD_CONNECTION.equals(command)) {
             return null;
         }
+        
+        String userId = null;
+        String dbUrl = null;
+        String driverClass = null;
+        
+        final Map m = gson.fromJson(gson.toJson(arguments.get(0)), Map.class);
+        if (m != null) {
+            userId = (String) m.get(USER_ID);
+            dbUrl = (String) m.get(DB_URL);
+            driverClass = (String) m.get(DRIVER);
+            
+        }
+        if (dbUrl != null && driverClass != null) {
+            
+            JDBCDriver[] driver = JDBCDriverManager.getDefault().getDrivers(driverClass); //NOI18N
+            if (driver != null && driver.length > 0) {
+                CompletableFuture<String> usernameFuture = userId != null ? CompletableFuture.completedFuture(userId) : client.showInputBox(new ShowInputBoxParams(
+                        Bundle.MSG_EnterUsername(), userId));
+                
+                usernameFuture.thenAccept((username) -> { //NOI18N
+                    if (username == null) {
+                        return;
+                    }
+                    char[] password = m.get(PASSWORD) == null ? null 
+                            : ((List<Double>) m.get(PASSWORD)).stream().map(n -> Character.toString((char) n.byteValue())).collect(Collectors.joining()).toCharArray();
+                    CompletableFuture<String> passwordFuture = password != null ? CompletableFuture.completedFuture(new String(password)) : client.showInputBox(new ShowInputBoxParams(
```

Comment:
Can't the sending side be expected to provide the password with the data instead of relying on additional callbacks? The same question applies to the username.

Reply:
This callbacks are used for user input in case sending side didn't provide username and password

Reply:
This does not answer the question why the sending side does not directly ask the user and provides everything necessary?

Reply:
All the user input for this command is in this particular command. Written in Java. I don't see any sense to split this command and implement part of it in typescript.

Reply:
I'm not sure that passing username and password as the command arguments over LSP is a good idea.